InnerEye/ML/SSL/ssl_online_evaluator.py

```diff
@@ -0,0 +1,178 @@
+        """
+        Get and log validation metrics.
+        """
+        ids_linear_head = tuple(batch[SSLModule.LINEAR_HEAD][0].tolist())
+        if ids_linear_head not in self.visited_ids:
+            self.visited_ids.add(ids_linear_head)
+            loss = self.shared_step(batch, pl_module, is_training=False)
+            pl_module.log('ssl/online_val_loss', loss, on_step=False, on_epoch=True, sync_dist=False)
+            for metric in self.val_metrics:
+                pl_module.log(f"ssl/online_val_{metric.name}", metric, on_epoch=True, on_step=False)  # type: ignore
+
+    def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx) -> None:  # type: ignore
+        """
+        Get and log training metrics, perform network update.
+        """
+        ids_linear_head = tuple(batch[SSLModule.LINEAR_HEAD][0].tolist())
+        if ids_linear_head not in self.visited_ids:
+            self.visited_ids.add(ids_linear_head)
+            loss = self.shared_step(batch, pl_module, is_training=True)
+            loss.backward()
+            self.optimizer.step()
+            self.optimizer.zero_grad()
+            self.training_step += 1
+
+            # log metrics
+            pl_module.log('ssl/online_train_loss', loss)
+            for metric in self.train_metrics:
+                pl_module.log(f"ssl/online_train_{metric.name}", metric, on_epoch=True, on_step=False)  # type: ignore
+
+
+def get_encoder_output_dim(pl_module: Union[pl.LightningModule, torch.nn.Module],
+                           dm: Optional[pl.LightningDataModule] = None) -> int:
+    """
+    Calculates the output dimension of ssl encoder by making a single forward pass.
+    :param pl_module: pl encoder module
+    :param dm: pl datamodule
+    """
+    # Target device
+    device = pl_module.device if isinstance(pl_module, pl.LightningDataModule) else \
+        next(pl_module.parameters()).device  # type: ignore
+    assert (isinstance(device, torch.device))
+
+    # Create a dummy input image
+    if dm is not None:
+        dataloader = dm.train_dataloader()
+        dataloader = dataloader[SSLModule.LINEAR_HEAD] if isinstance(dataloader, dict) else dataloader
+        batch = iter(dataloader).next()  # type: ignore 
+        x, _ = SSLOnlineEvaluatorInnerEye.to_device(batch, device)
+    else:
+        x = torch.rand((1, 3, 256, 256)).to(device)
+
+    # Extract the number of output feature dimensions
+    with torch.no_grad():
+        representations = pl_module(x)
+
+    return representations.shape[1]
+
+
+def WrapSSL(ssl_class: Any, num_classes: int) -> Any:
+    """
+    Wraps a given SSL encoder and adds a non-linear evaluator to it. This is done to load pre-trained SSL checkpoints.
+    PL requires non_linear_evaluator to be included in pl_module at SSL training time.
+    :param num_classes: Number of target classes for the linear head.
+    :param ssl_class:   SSL object either BYOL or SimCLR.
+    """
```

Comment:
So basically we cannot just do BYOLInnerEye.load("ckpt") with strict loading because the checkpoint will contain the weights of the linear evaluator, but this one is defined outside of the BYOLInnerEye class, hence we can only load the checkpoint if we manually re-add the linear evaluator prior to loading. 
Will add this comment to the doc and renaming the class to `SSLModelLoader`